### PR TITLE
add op npx.index_update

### PR DESCRIPTION
**Pull request by JiangZhaoh**
Add npx op index_update.
Relate to #17823 .

### CODE REVIEW

**Comment by JiangZhaoh** [comment]
@mxnet-bot run ci [unix-gpu, unix-cpu, windows-cpu]

File: python/mxnet/_numpy_op_doc.py [hunk 1]
```diff
@@ -699,6 +699,71 @@ def _npx_index_add(a, ind, val):
     pass
 
 
+def _npx_index_update(a, ind, val):
+    """
+    Update values to input according to given indexes.
+    If multiple indices refer to the same location it is undefined which update is chosen; it may choose
+    the order of updates arbitrarily and nondeterministically (e.g., due to concurrent updates on some
+    hardware platforms). Recommend not to use repeate positions.
```
Comment (sxjscience) [inline python/mxnet/_numpy_op_doc.py:707]: You may add a blank line here. (I find that `index_add` misses the blank line also).

File: src/operator/tensor/index_update.cu [hunk 2]
```diff
@@ -0,0 +1,202 @@
...
+                                     const TBlob& ind,
+                                     const mshadow::Shape<MXNET_SPECIAL_MAX_NDIM> grada_pre_stride,
+                                     const int tail_size, const int ind_num, const int ind_ndim,
+                                     const int seg, const int req) {
+  using namespace mxnet_op;
+  using namespace mshadow;
+  mshadow::Stream<gpu> *s = ctx.get_stream<gpu>();
+  MSHADOW_TYPE_SWITCH(grad_a.type_flag_, DType, {
+    size_t temp_mem_size = ind.shape_.Size() * sizeof(int) +
+                           ograd.shape_.Size() * sizeof(DType);
```
Comment (sxjscience) [inline src/operator/tensor/index_update.cu:174]: Here, we mix `int` type and `DType`. The second pointer may not be appropriately aligned with `DType`. You may consider to revise the code and add the `PadBytes`: https://github.com/apache/incubator-mxnet/blob/1b02225fefd8ccc93bc73223f0d3cde103fad661/src/operator/tensor/ordering_op-inl.h#L350-L351

The reason to do this is that CUDA forces memory alignment, i.e., we will always have the following assertion:

```c++
DType* ptr;
ASSERT static_cast<size_t>(ptr) % sizeof(DType) == 0.
```

By default, calling cudaMalloc will give you a pointer with at least 256-bytes aligned address. However, after shifting the pointer to `ptr += ind.shape_.Size() * sizeof(int)`, the new address may not be aligned with `sizeof(DType)`.

File: tests/python/unittest/test_numpy_op.py [hunk 3]
```diff
@@ -1477,6 +1477,165 @@ def index_add_bwd(out_grad, a_grad, ind, val_grad, ind_ndim, ind_num, grad_req_a
...
+            configs.append(tuple([shape, ind, tuple(val_shape), ind_ndim, ind_num]))
+
+    dtypes = ['float32', 'float64', 'int32', 'int64']
+    grad_req = ['write', 'null', 'add']
+    for hybridize, grad_req_a, grad_req_val, dtype, indtype in \
+        itertools.product([True, False], grad_req, grad_req, dtypes, ['int32', 'int64']):
+        for a_shape, ind, val_shape ,ind_ndim, ind_num in configs:
+            eps = 1e-3
+            if sys.platform.startswith('linux'):
+                eps = 1e-2
```
Comment (leezu) [inline tests/python/unittest/test_numpy_op.py:1589]: eps should not depend on the platform. What's the root-cause here?

File: src/operator/tensor/index_update.cu (hunk 2 above)
Comment (JiangZhaoh) [reply src/operator/tensor/index_update.cu]: OK, I have revised this segment in [https://github.com/apache/incubator-mxnet/pull/18545/commits/c251837d2703813a254fd75cc426d50139220272](url)

Comment (JiangZhaoh) [reply src/operator/tensor/index_update.cu]: OK, I have revised the code in commit [https://github.com/apache/incubator-mxnet/pull/18545/commits/c251837d2703813a254fd75cc426d50139220272](url)

**Comment by JiangZhaoh** [comment]
@mxnet-bot run ci [unix-cpu]

**Pull request merged**